File: ts-src/objects/GameState.ts
import { Board } from "./Board";
import { Player } from "./Player";

interface IGameState {
  players?: Player[];
  board?: Board;
  score?: number;
  rotations?: number;
  rotateAngle?: number;
  outcome?: string;
}

export class GameState {
  public players: Player[] = [];
  public board: Board = null;
  public score: number = 0;
  public rotations: number = 0;
  public rotateAngle: number = 0;
  public outcome: string = "";

  constructor(params: IGameState = {}) {
    (Object as any).entries(params).map(pair => {
      const [key, value] = pair;
      this[key] = value;
    });
  }

  public modify(values: IGameState = {}) {
    return new GameState((Object as any).assign({}, this, values));
  }
}

File: rust/src/types/game_state.rs
use super::action::Action;
use super::board::Board;
use super::outcome::Outcome;
use super::player::Player;
use super::render_angle::RenderAngle;
use super::score::Score;
use super::screen_size::ScreenSize;

#[derive(Clone, Debug)]
pub struct GameState {
    pub players: Vec<Player>,
    pub board: Board,
    pub score: Score,
    pub rotations: i32,
    pub rotate_angle: RenderAngle,
    pub render_angle: RenderAngle,
    pub outcome: Outcome,
    pub turns: i32,
    pub current: Action,
    pub screen_size: ScreenSize,
}

impl GameState {
    pub fn new(board: Board) -> Self {
        GameState {
            players: Vec::new(),
            board,
            score: Score(0),
            rotations: 0,
            rotate_angle: RenderAngle(0),
            render_angle: RenderAngle(0),
            outcome: Outcome::KeepPlaying,
            turns: 0,
            current: Action::Playing,
            screen_size: ScreenSize::new(1, 1),
        }
    }
}
